stl/inc/functional

```diff
@@ -795,7 +795,8 @@ private:
 constexpr size_t _Space_size = (_Small_object_num_ptrs - 1) * sizeof(void*);
 
 template <class _Impl> // determine whether _Impl must be dynamically allocated
-_INLINE_VAR constexpr bool _Is_large = (_Space_size < sizeof(_Impl)) || !_Impl::_Nothrow_move::value;
+_INLINE_VAR constexpr bool _Is_large = (_Space_size < sizeof(_Impl)) || alignof(_Impl) > alignof(max_align_t)

```

Comment:
For other reviewers: `_Is_large` is used in member functions of `_Func_impl`, `_Func_impl_no_alloc`, and `_Func_class`. The prior two are used by `_Func_class`, and `function` derives from `_Func_class`. The value of `_Is_large` has changed for small, nothrow-movable, highly-aligned types. Consequently there could be ODR violations if new code that puts such a type `T` into `std::function` is linked with old objects that contain instantiations of the members of any of the above class templates for `T`.

We could address these potential ODR violations by changing the names of the above three class templates, but given that the potentially problematic old code was broken (that's the bug we're fixing here), I speculate there are very few or no occurrences in-the-wild. Addressing the ODR violations seems unnecessary to me.

If anyone else assesses the risk differently - I should note here that I don't think there's any risk of problems if we do change the names of the three class templates - please speak up.

Reply:
> We could address these potential ODR violations by changing the names of the above three class templates

I'm afraid then `function` itself is fourth template to rename.

So there's no option to rename I think. We probably only have `<atomic>` approach to yield the responsibility for breaking ODR to the user:

https://github.com/microsoft/STL/blob/22391019cba3863a978ea2d7d3f5ca240f38c576/stl/inc/atomic#L1508-L1516

Reply:
> I'm afraid then `function` itself is fourth template to rename.

For `function` the only problematic members are those that know the type of a stored function object: the converting constructor template and converting assignment operator templates. We could "rename" those by adding defaulted template parameters. Other members are insulated by type erasure.

But like I said, I don't think there's sufficient motivation to bother.

Reply:
Agreed. We escape-hatched atomic because in theory, someone could have manually aligned the location. But `std::function` is copyable, so even if someone manually aligned a particular `std::function` storing a highly-aligned function object locally, any copies wouldn't necessarily have received such treatment.

Therefore, all old object files triggering this scenario were extremely likely to have had silent bad codegen, and this ODR issue can't make silent bad codegen worse.